File: PaintHouseLeetcode256.java
import java.util.Arrays;

/* Solution 1 Recursion */
//Time Complexity: O(2^N) permutations, where is the  N is the level of the recursion. At every leafs we will have 2 choices of color for each house.
//Leet code : yes
//Space Complexity : O(2^N) //level of dpeth of recursion tree.
public class PaintHouseLeetcode256 {
        int min_cost = 999;
        public int minCost(int[][] costs) {

            if (costs == null) {
                return 0;
            }

            return helper(costs, 0, -1);

        }

        public int helper(int[][]costs, int r,int lc) {

            if (r >= costs.length) {
                return 0;
            }

            for (int i = 0; i < costs[0].length; i++) {
                //lc is last color by this we know which color was choosen fo previous so we can skip this.
                if (i == lc) continue;
                System.out.println("Before calling helper r : "+r+" i : "+i+" lc : "+lc+" cost : "+ min_cost);
                min_cost = Math.min(min_cost, costs[r][i] + helper(costs, r + 1, i));
                System.out.println("After calling helper  r : "+r+" i : "+i+" lc : "+lc+" cost : "+ min_cost);


            }

            return min_cost;
        }

        /* Solution 2 Memoization */
//Time Complexity: O(N)  where is the  N is the level of recursion or the number of houses..we will calculate each color * each house. (less no of colors we drop it and say O(N)
//Leet code : yes
//Space Complexity : O(N) //level of dpeth of recursion tree.

        int dp[][];
        public int minCost1(int[][] costs) {
             dp = new int[costs.length + 1][3];
            for(int row[]: dp){
                Arrays.fill(row, -1);
            }
            return helper1(costs, costs.length, -1);
        }

        public int helper1(int[][] costs, int n, int prev) {
            if(n == 0){
                return 0;
            }

            int ans = Integer.MAX_VALUE;
            for(int i = 0; i < 3; i++){
                if(i == prev){
                    continue;
                }
                if(dp[n][i] != -1){
                    ans = Math.min(ans, dp[n][i]);
                }else{
                    int result = costs[n-1][i] + helper1(costs, n - 1, i);
                    dp[n][i] = result;
                    ans = Math.min(ans, result);
                }


            }

            return ans;

        }


}

File: Paint_House_Leetcode256.py
class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        # SC - O(1) DP without extra space
        # TC - O(n)
        n = len(costs)
        colorR = costs[n - 1][0]
        colorB = costs[n - 1][1]
        colorG = costs[n - 1][2]

        for i in range(n - 2, -1, -1):
            TempR = colorR
            TempB = colorB
            colorR = costs[i][0] + min(colorB, colorG)
            colorB = costs[i][1] + min(TempR, colorG)
            colorG = costs[i][2] + min(TempR, TempB)

        return min(colorR, colorB, colorG)

    '''
        # DP without altering costs thereby using SC = O(3*n) = O(n)
        # TC - O(n)

        n = len(costs)
        dp = [[0]*3 for _ in range(n)]
        dp[n-1] = costs[n-1] #intialize last row of dp same as costs last row

        for i in range(n-2,-1,-1):
            dp[i][0] =costs[i][0] + min(dp[i+1][1], dp[i+1][2])
            dp[i][1] =costs[i][1] + min(dp[i+1][0], dp[i+1][2])
            dp[i][2] =costs[i][2] + min(dp[i+1][0], dp[i+1][1])

        return min(dp[0][0],dp[0][1],dp[0][2])
    '''

    '''
        # DP using the same costs matrix
        # TC - O(n)
        # SC - O(1)

        n = len(costs)
        for i in range(n-2,-1,-1):
            costs[i][0] =costs[i][0] + min(costs[i+1][1], costs[i+1][2])
            costs[i][1] =costs[i][1] + min(costs[i+1][0], costs[i+1][2])
            costs[i][2] =costs[i][2] + min(costs[i+1][0], costs[i+1][1])

        return min(costs[0][0],costs[0][1],costs[0][2])
    '''
    '''
        # Recursion Time Limit exceeded
        # TC O(3*(2**n))
        # SC O(n)
        color_0 = self.helper(costs, 0 , 0, 0)
        color_1 = self.helper(costs, 0 , 1, 0)
        color_2 = self.helper(costs, 0 , 2, 0)

        return min(color_0, color_1,color_2)

    def helper(self, costs, i , color, mincost):
        #base
        if i == len(costs):
            return mincost

        #logic
        if color == 0:
            return min(self.helper(costs, i+1 , 1, mincost + costs[i][0]), self.helper(costs, i+1,2, mincost + costs[i][0]))

        elif color == 1:
            return min(self.helper(costs, i+1, 2, mincost + costs[i][1]), self.helper(costs, i+1 , 0, mincost + costs[i][1]))

        elif color == 2:
            return min(self.helper(costs, i+1 , 1, mincost + costs[i][2]),self.helper(costs, i+1 ,0, mincost + costs[i][2]))
    '''
